# Домашняя работа 2

### Условие 1: 1 задача

Скачать данные по ссылке https://www.kaggle.com/datasets/shivachandel/kc-house-data
Считать данные с помощью pandas.
Вывести на экран первые 5 строк.
Посмотреть на описание признаков и на их содержание.

In [1]:
import pandas as pd

Считать данные с помощью pandas.

In [12]:
df = pd.read_csv('Курс Pyhton для разработчиков/kc_house_data.csv', sep = ',')
df

id             date     price  bedrooms  bathrooms  \
0      7129300520  20141013T000000  221900.0         3       1.00   
1      6414100192  20141209T000000  538000.0         3       2.25   
2      5631500400  20150225T000000  180000.0         2       1.00   
3      2487200875  20141209T000000  604000.0         4       3.00   
4      1954400510  20150218T000000  510000.0         3       2.00   
...           ...              ...       ...       ...        ...   
21608   263000018  20140521T000000  360000.0         3       2.50   
21609  6600060120  20150223T000000  400000.0         4       2.50   
21610  1523300141  20140623T000000  402101.0         2       0.75   
21611   291310100  20150116T000000  400000.0         3       2.50   
21612  1523300157  20141015T000000  325000.0         2       0.75   

       sqft_living  sqft_lot  floors  waterfront  view  ...  grade  \
0             1180      5650     1.0           0     0  ...      7   
1             2570      7242     2.0           0     0  ...      7   
2              770     10000     1.0           0     0  ...      6   
3             1960      5000     1.0           0     0  ...      7   
4             1680      8080     1.0           0     0  ...      8   
...            ...       ...     ...         ...   ...  ...    ...   
21608         1530      1131     3.0           0     0  ...      8   
21609         2310      5813     2.0           0     0  ...      8   
21610         1020      1350     2.0           0     0  ...      7   
21611         1600      2388     2.0           0     0  ...      8   
21612         1020      1076     2.0           0     0  ...      7   

       sqft_above  sqft_basement  yr_built  yr_renovated  zipcode      lat  \
0          1180.0              0      1955             0    98178  47.5112   
1          2170.0            400      1951          1991    98125  47.7210   
2           770.0              0      1933             0    98028  47.7379   
3          1050.0            910      1965             0    98136  47.5208   
4          1680.0              0      1987             0    98074  47.6168   
...           ...            ...       ...           ...      ...      ...   
21608      1530.0              0      2009             0    98103  47.6993   
21609      2310.0              0      2014             0    98146  47.5107   
21610      1020.0              0      2009             0    98144  47.5944   
21611      1600.0              0      2004             0    98027  47.5345   
21612      1020.0              0      2008             0    98144  47.5941   

          long  sqft_living15  sqft_lot15  
0     -122.257           1340        5650  
1     -122.319           1690        7639  
2     -122.233           2720        8062  
3     -122.393           1360        5000  
4     -122.045           1800        7503  
...        ...            ...         ...  
21608 -122.346           1530        1509  
21609 -122.362           1830        7200  
21610 -122.299           1020        2007  
21611 -122.069           1410        1287  
21612 -122.299           1020        1357  

[21613 rows x 21 columns]

Вывести на экран первые 5 строк.

In [13]:
df.head()

id             date     price  bedrooms  bathrooms  sqft_living  \
0  7129300520  20141013T000000  221900.0         3       1.00         1180   
1  6414100192  20141209T000000  538000.0         3       2.25         2570   
2  5631500400  20150225T000000  180000.0         2       1.00          770   
3  2487200875  20141209T000000  604000.0         4       3.00         1960   
4  1954400510  20150218T000000  510000.0         3       2.00         1680   

   sqft_lot  floors  waterfront  view  ...  grade  sqft_above  sqft_basement  \
0      5650     1.0           0     0  ...      7      1180.0              0   
1      7242     2.0           0     0  ...      7      2170.0            400   
2     10000     1.0           0     0  ...      6       770.0              0   
3      5000     1.0           0     0  ...      7      1050.0            910   
4      8080     1.0           0     0  ...      8      1680.0              0   

   yr_built  yr_renovated  zipcode      lat     long  sqft_living15  \
0      1955             0    98178  47.5112 -122.257           1340   
1      1951          1991    98125  47.7210 -122.319           1690   
2      1933             0    98028  47.7379 -122.233           2720   
3      1965             0    98136  47.5208 -122.393           1360   
4      1987             0    98074  47.6168 -122.045           1800   

   sqft_lot15  
0        5650  
1        7639  
2        8062  
3        5000  
4        7503  

[5 rows x 21 columns]

Посмотреть на описание признаков и на их содержание.

In [14]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21613 entries, 0 to 21612
Data columns (total 21 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   id             21613 non-null  int64  
 1   date           21613 non-null  object 
 2   price          21613 non-null  float64
 3   bedrooms       21613 non-null  int64  
 4   bathrooms      21613 non-null  float64
 5   sqft_living    21613 non-null  int64  
 6   sqft_lot       21613 non-null  int64  
 7   floors         21613 non-null  float64
 8   waterfront     21613 non-null  int64  
 9   view           21613 non-null  int64  
 10  condition      21613 non-null  int64  
 11  grade          21613 non-null  int64  
 12  sqft_above     21611 non-null  float64
 13  sqft_basement  21613 non-null  int64  
 14  yr_built       21613 non-null  int64  
 15  yr_renovated   21613 non-null  int64  
 16  zipcode        21613 non-null  int64  
 17  lat            21613 non-null  float64
 18  long  

### Условие 2: 2 задача
Проведите первичный анализ данных.
Изучите типы данных.
Найдите количество пропущенных ячеек в данных.
Посчитайте основные статистики по всем признакам и поизучайте их.
Пишите выводы.

Изучите типы данных. Найдите количество пропущенных ячеек в данных.

In [30]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21613 entries, 0 to 21612
Data columns (total 21 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   id             21613 non-null  int64  
 1   date           21613 non-null  object 
 2   price          21613 non-null  float64
 3   bedrooms       21613 non-null  int64  
 4   bathrooms      21613 non-null  float64
 5   sqft_living    21613 non-null  int64  
 6   sqft_lot       21613 non-null  int64  
 7   floors         21613 non-null  float64
 8   waterfront     21613 non-null  int64  
 9   view           21613 non-null  int64  
 10  condition      21613 non-null  int64  
 11  grade          21613 non-null  int64  
 12  sqft_above     21611 non-null  float64
 13  sqft_basement  21613 non-null  int64  
 14  yr_built       21613 non-null  int64  
 15  yr_renovated   21613 non-null  int64  
 16  zipcode        21613 non-null  int64  
 17  lat            21613 non-null  float64
 18  long  

Вывод:
типы данных - с плавающей запятой 6 полей, целые числа 14 полей, объект 1 поле.
пропущенных ячеек нет, все данные при подсчете 21613 и с отметкой "non-null"

Посчитайте основные статистики по всем признакам и поизучайте их. Пишите выводы.

In [15]:
df.describe()

id         price      bedrooms     bathrooms   sqft_living  \
count  2.161300e+04  2.161300e+04  21613.000000  21613.000000  21613.000000   
mean   4.580302e+09  5.400881e+05      3.370842      2.114757   2079.899736   
std    2.876566e+09  3.671272e+05      0.930062      0.770163    918.440897   
min    1.000102e+06  7.500000e+04      0.000000      0.000000    290.000000   
25%    2.123049e+09  3.219500e+05      3.000000      1.750000   1427.000000   
50%    3.904930e+09  4.500000e+05      3.000000      2.250000   1910.000000   
75%    7.308900e+09  6.450000e+05      4.000000      2.500000   2550.000000   
max    9.900000e+09  7.700000e+06     33.000000      8.000000  13540.000000   

           sqft_lot        floors    waterfront          view     condition  \
count  2.161300e+04  21613.000000  21613.000000  21613.000000  21613.000000   
mean   1.510697e+04      1.494309      0.007542      0.234303      3.409430   
std    4.142051e+04      0.539989      0.086517      0.766318      0.650743   
min    5.200000e+02      1.000000      0.000000      0.000000      1.000000   
25%    5.040000e+03      1.000000      0.000000      0.000000      3.000000   
50%    7.618000e+03      1.500000      0.000000      0.000000      3.000000   
75%    1.068800e+04      2.000000      0.000000      0.000000      4.000000   
max    1.651359e+06      3.500000      1.000000      4.000000      5.000000   

              grade    sqft_above  sqft_basement      yr_built  yr_renovated  \
count  21613.000000  21611.000000   21613.000000  21613.000000  21613.000000   
mean       7.656873   1788.396095     291.509045   1971.005136     84.402258   
std        1.175459    828.128162     442.575043     29.373411    401.679240   
min        1.000000    290.000000       0.000000   1900.000000      0.000000   
25%        7.000000   1190.000000       0.000000   1951.000000      0.000000   
50%        7.000000   1560.000000       0.000000   1975.000000      0.000000   
75%        8.000000   2210.000000     560.000000   1997.000000      0.000000   
max       13.000000   9410.000000    4820.000000   2015.000000   2015.000000   

            zipcode           lat          long  sqft_living15     sqft_lot15  
count  21613.000000  21613.000000  21613.000000   21613.000000   21613.000000  
mean   98077.939805     47.560053   -122.213896    1986.552492   12768.455652  
std       53.505026      0.138564      0.140828     685.391304   27304.179631  
min    98001.000000     47.155900   -122.519000     399.000000     651.000000  
25%    98033.000000     47.471000   -122.328000    1490.000000    5100.000000  
50%    98065.000000     47.571800   -122.230000    1840.000000    7620.000000  
75%    98118.000000     47.678000   -122.125000    2360.000000   10083.000000  
max    98199.000000     47.777600   -121.315000    6210.000000  871200.000000

Вывод:
Количество записей 21 613
выведены средние значения по каждому столбцу, среднее квадратическое отклонение, минимальные и максимальные значения, и квантили 25%, 50%, 75%.

In [34]:
df.describe(include = ['object'])

date
count             21613
unique              372
top     20140623T000000
freq                142

Вывод: с параметром object выдана статистика только по полю date

### Условие 3: 3 задача
Ответьте на несколько вопросов
3.1 В каком диапазоне изменяются стоимости недвижимости?
3.2 Какую долю в среднем занимают жилая площадь от всей площади по всем домам?
3.3 Как много домов с разными этажами в данных?
3.4 Насколько хорошие состояния у домов в данных?
3.5 Найдите года, когда построили первый дом, когда построили последний дом в данных?

3.1 В каком диапазоне изменяются стоимости недвижимости?

Стоимость недвижимости изменяется в диапазоне согласно данных статистике, выведенной во втором задании, от 75 000 до 7 700 000.

In [43]:
# или вывести отдельной статистикой по полю price
df['price'].describe()

count    2.161300e+04
mean     5.400881e+05
std      3.671272e+05
min      7.500000e+04
25%      3.219500e+05
50%      4.500000e+05
75%      6.450000e+05
max      7.700000e+06
Name: price, dtype: float64

In [44]:
df['price'].min()

75000.0

In [45]:
df['price'].max()

7700000.0

3.2 Какую долю в среднем занимают жилая площадь от всей площади по всем домам?

In [32]:
# средняя площадь проживания
sqft_living = 2079.899736
# средняя общая площадь
sqft_lot = 15106.97
print(f'Доля средней жилой площади в средней площади по всем домам {sqft_living/sqft_lot} или {100*sqft_living/sqft_lot} %')

Доля средней жилой площади в средней площади по всем домам 0.1376781535940033 или 13.76781535940033 %


In [46]:
# средняя по жилой площади
df['sqft_living'].mean()

2079.8997362698374

In [47]:
# средняя по общей площади
df['sqft_lot'].mean()

15106.967565816869

In [48]:
avg = df['sqft_living'].mean() / df['sqft_lot'].mean()

In [52]:
print(f'Доля средней жилой площади в средней площади по всем домам {avg} или {avg*100} %')

Доля средней жилой площади в средней площади по всем домам 0.1376781757959227 или 13.76781757959227 %


Вывод: С использованием метода df['sqft_lot'].mean() дает более точный расчет от метода ручной записи из данных общей статистической таблицы

In [53]:
print(f'Доля средней жилой площади в средней площади по всем домам {round(avg,4)} или {round(avg*100,2)} %')

Доля средней жилой площади в средней площади по всем домам 0.1377 или 13.77 %


3.3 Как много домов с разными этажами в данных?

In [65]:
df['floors'].value_counts()

1.0    10680
2.0     8241
1.5     1910
3.0      613
2.5      161
3.5        8
Name: floors, dtype: int64

3.4 Насколько хорошие состояния у домов в данных?

In [59]:
# максимальное значение в поле grade
mx_grade = df['grade'].max()
mx_grade

13

In [57]:
df['grade'].value_counts()

7     8981
8     6068
9     2615
6     2038
10    1134
11     399
5      242
12      90
4       29
13      13
3        3
1        1
Name: grade, dtype: int64

In [60]:
df['grade'].describe(include = ['object'])

count    21613.000000
mean         7.656873
std          1.175459
min          1.000000
25%          7.000000
50%          7.000000
75%          8.000000
max         13.000000
Name: grade, dtype: float64

In [ ]:
Вывод: больше всего домов с оценкой 7, на втором месте с оценкой 8, на третьем месте с оценкой 9.

3.5 Найдите года, когда построили первый дом, когда построили последний дом в данных?

In [54]:
df['yr_built'].min()

1900

In [55]:
df['yr_built'].max()

2015

Вывод: самый ранний год постройки дома - 1900, дом с самым поздним годом постройки - 2015

### Условие 4: 4 задача
Ответьте на несколько вопросов

4.1 Сколько в среднем стоят дома, у которых 2 спальни?
4.2 Какая в среднем общая площадь домов, у которых стоимость больше 600 000?
4.3 Как много домов коснулся ремонт?
4.4 Насколько в среднем стоимость домов с оценкой grade домов выше 10 отличается от стоимости домов с оценкой grade меньше 4?

4.1 Сколько в среднем стоят дома, у которых 2 спальни?

In [72]:
# новый дата фрейм с домами 2 спальни
df1 = df[(df['bedrooms'] == 2)]

In [77]:
# средняя в новом дата фрейме по полю цена
avg_2b = df1['price'].mean()

In [80]:
print(f'Средняя цена дома с двумя спальнями {round(avg_2b,2)}')

Средняя цена дома с двумя спальнями 401372.68


4.2 Какая в среднем общая площадь домов, у которых стоимость больше 600 000?

In [81]:
df2 = df[(df['price'] > 600000)]

In [84]:
avg_u6 = df2['sqft_lot'].mean()

In [86]:
print(f'Средняя общая площадь домов дороже 600 000 = {round(avg_u6,2)}')

Средняя общая площадь домов дороже 600 000 = 20442.52


 4.3 Как много домов коснулся ремонт? 

In [126]:
df3 = df[(df['yr_renovated'] > 0)]

In [127]:
cnt = df3['yr_renovated'].count()

In [128]:
print(f'Количество домов, которые были отремонтированы = {cnt} домов')

Количество домов, которые были отремонтированы = 914 домов


In [130]:
# второй способ
cnt1 = df[(df['yr_renovated'] > 0)]['yr_renovated'].count()
print(f'Количество домов, которые были отремонтированы = {cnt1} домов')

Количество домов, которые были отремонтированы = 914 домов


4.4 Насколько в среднем стоимость домов с оценкой grade домов выше 10 отличается от стоимости домов с оценкой grade меньше 4?

In [115]:
df_g10 = df[(df['grade'] > 10)]

In [116]:
df_g4 = df[(df['grade'] < 4)]

In [121]:
price10 = df_g10['price'].mean()

In [122]:
price4 = df_g4['price'].mean()

In [123]:
value = round(price10 - price4,2)

In [124]:
print(f'Средняя цена домов с оценкой выше 10 отличается на {value} от средней цены домов с оценкой меньше 4')

Средняя цена домов с оценкой выше 10 отличается на 1488885.12 от средней цены домов с оценкой меньше 4


### Условие 5: 5 задача
Ответьте на несколько вопросов
5.1 Выберите дом клиенту
Клиент хочет дом с видом на набережную, как минимум с тремя ванными и с подвалом. Сколько вариантов есть у клиента?
5.2 Выберите дом клиенту
Клиент хочет дом либо с очень красивым видом из окна, либо с видом на набережную, в очень хорошем состоянии и год постройки не меньше 1980 года. В какой ценовом диапазоне будут дома?
5.3 Выберите дом клиенту
Клиент хочет дом без подвала, с двумя этажами, стоимостью до 150000. Какая оценка по состоянию у таких домов в среднем?

5.1 Выберите дом клиенту. Клиент хочет дом с видом на набережную, как минимум с тремя ванными и с подвалом. Сколько вариантов есть у клиента?

In [145]:
# первый вариант
df5_1 = df[(df['bathrooms'] >=3)]
df5_11 = df5_1[(df5_1['sqft_basement'] > 0)]
df5_111 = df5_11[(df5_11['waterfront'] == 1)]

In [188]:
# второй вариант - вложенные условия
df5_12 = df[(df['bathrooms'] >=3) & (df['sqft_basement'] > 0) & (df['waterfront'] == 1)]
cnt5_12 = df5_12['date'].count()

In [189]:
print(f'Количество ваннами не менее 3, подвалом и видом на набережную {cnt5_12}')

Количество ваннами не менее 3, подвалом и видом на набережную 41


5.2 Выберите дом клиенту. Клиент хочет дом либо с очень красивым видом из окна, либо с видом на набережную, в очень хорошем состоянии и год постройки не меньше 1980 года. В какой ценовом диапазоне будут дома?

In [190]:
df5_2 = df[(df['view'] == df['view'].max()) | (df['waterfront'] == 1)]

In [191]:
df5_21 = df5_2[(df5_2['yr_built'] >= 1980) & (df5_2['grade'] == df5_2['grade'].max())]

In [192]:
cmn = df5_21['price'].min()

In [193]:
cmx = df5_21['price'].max()

In [194]:
c5_2 = df5_21['id'].count()

In [195]:
print(f'Стоимость дома от {cmn} до {cmx}. Всего {c5_2} вариантов')

Стоимость дома от 3800000.0 до 6885000.0. Всего 2 вариантов


5.3 Выберите дом клиенту. Клиент хочет дом без подвала, с двумя этажами, стоимостью до 150000. Какая оценка по состоянию у таких домов в среднем?

In [198]:
df5_3 = df[(df['floors'] == 2) & (df['sqft_basement'] == 0) & (df['price'] < 150000)]
df5_3

id             date     price  bedrooms  bathrooms  \
2589   5061300030  20140508T000000  134000.0         2       1.50   
10105  5466310060  20150324T000000  139500.0         2       1.50   
15610  7614100080  20150211T000000  140000.0         3       1.75   
18332  9272202260  20140924T000000  130000.0         3       1.00   
18939  4239400300  20141129T000000   90000.0         3       1.00   
18988  7399100210  20141126T000000  140000.0         3       1.50   

       sqft_living  sqft_lot  floors  waterfront  view  ...  grade  \
2589           980      5000     2.0           0     0  ...      7   
10105         1230      1561     2.0           0     0  ...      7   
15610         1270      8991     2.0           0     0  ...      7   
18332         1200      7000     2.0           0     0  ...      7   
18939          980      2490     2.0           0     0  ...      6   
18988         1200      2002     2.0           0     0  ...      8   

       sqft_above  sqft_basement  yr_built  yr_renovated  zipcode      lat  \
2589        980.0              0      1922          2003    98014  47.7076   
10105      1230.0              0      1983             0    98042  47.3768   
15610      1270.0              0      1981             0    98042  47.3563   
18332      1200.0              0      1908             0    98116  47.5883   
18939       980.0              0      1969             0    98092  47.3170   
18988      1200.0              0      1966             0    98055  47.4659   

          long  sqft_living15  sqft_lot15  
2589  -121.359           1040        5000  
10105 -122.149           1660        2243  
15610 -122.149           1270        8993  
18332 -122.384           3290        6000  
18939 -122.182            980        3154  
18988 -122.189           1270        1848  

[6 rows x 21 columns]

In [201]:
avg_g = df5_3['grade'].mean()

In [202]:
print(f'Средняя оценка домов до 150000 с подвалом и 2 этажа = {avg_g}')

Средняя оценка домов до 150000 с подвалом и 2 этажа = 7.0
